### PR TITLE
MM-33993 Allow apps form and interactive dialogs to clear select fields

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

This PR makes it so user, channel, and option select boxes can be cleared.

Note that this functionality is not introduced for interactive messages or embedded forms.

#### Ticket Link

Fixes https://mattermost.atlassian.net/browse/MM-33993

#### Screenshots

Android Pixel 2

#### Device Info

Android Pixel 2

iPhone simulator 12.3

#### Release Note

```release-note
Select boxes in app forms and interactive dialogs can be cleared now
```

### CODE REVIEW

**Comment by matthewbirtch** [comment]
@mickmister what's the best way for me to test this from a UX perspective? Will I be able to access the interactions if I start a build?

**Review by larkox — APPROVED** [review]

**Comment by matthewbirtch** [comment]
Now that I'm testing it in a device, I wonder if treating this input like a `select` input is the right way to go. It's unusual for a `select` input to push to a new screen. I don't think we should make it look like a select input if it's not opening the expected `picker` UI. Removing the dropdown arrow could likely be enough. But, if it's going to push a new screen into the stack, I would more expect there to be a chevron pointing to the right rather than down.

Maybe we consider two possibilities?
1) standard select/picker input that looks more like a standard dropdown and opens up simple text-only options in a picker UI
2) A row item that will transitions to a new screen to choose more rich content options (like users from a list with search UI for example)

Also, for the clear `x` interaction, it can be quite tough to tap it given the small area if occupies. Can we enlarge the tap area even more? (Minimum 40x40 area).

**Comment by mickmister** [comment]
> A row item that will transitions to a new screen to choose more rich content options (like users from a list with search UI for example)

@matthewbirtch I think this would be the better path, but I'm not sure exactly what you mean. What needs to change from the current implementation to get to your description here? I think the second screen currently implemented fits the description of "a new screen to choose more rich content options (like users from a list with search UI for example)".

> Also, for the clear `x` interaction, it can be quite tough to tap it given the small area if occupies. Can we enlarge the tap area even more? (Minimum 40x40 area).

:+1:

**Comment by matthewbirtch** [comment]
> I think the second screen currently implemented fits the description of "a new screen to choose more rich content options (like users from a list with search UI for example)"

It does fit this description. However, the input that triggers it probably shouldn't look like a select input. I think if you just remove the down arrow or change it to an arrow pointing to the right that might be enough

**Comment by mickmister** [comment]
@matthewbirtch I've increased the touch area of the x. It feels much better now.

Here's what it looks like with the "right chevron":




0/5 I personally think it looks better with the "down chevron". Even though it is navigating to another screen, it still feels like a dropdown select box to me.

**Comment by mickmister** [comment]
@enahum I tried your suggestion of using `hitSlop` to increase the "close x's" touch area, but it didn't work correctly because the touch area we are trying to increase overlaps with another touch area. Meaning the "close x" is on top of the select box. So I've gone the route of increasing the padding on the "close x" to increase its touchable area. Note that since it's enclosed in the select box view, the touchable area does not extend vertically past the height of the select box.

File: app/components/autocomplete_selector/autocomplete_selector.js [hunk 1]
```diff
@@ -198,10 +206,21 @@ export default class AutocompleteSelector extends PureComponent {
...
                         </Text>
+                        {!disabled && onClear && selected && (
+                            <CompassIcon
+                                name='close-circle'
+                                color={changeOpacity(theme.centerChannelColor, 0.5)}
+                                style={style.clearx}
+                                size={20}
+                                onPress={this.handleClear}
+                            />
+                        )}
```
Comment (enahum) [inline app/components/autocomplete_selector/autocomplete_selector.js:217]: @mickmister have you tried wrapping this with a [TapGestureHandler](https://docs.swmansion.com/react-native-gesture-handler/docs/api/gesture-handlers/tap-gh) or another Touchable?

Comment (mickmister) [reply app/components/autocomplete_selector/autocomplete_selector.js:217]: @enahum I have tried that with `TapGestureHandler` and `TouchableWithoutFeedback`, but `this.handleClear` is never called when I press on the button. There's no reaction to pressing the button. I've tried the following:

```js
<TouchableWithoutFeedback
    onPress={this.handleClear}
>
    <CompassIcon
        name='close-circle'
        color={changeOpacity(theme.centerChannelColor, 0.5)}
        style={style.clearx}
        size={20}
    />
</TouchableWithoutFeedback>


<TouchableWithoutFeedback
    onPress={this.handleClear}
>
    <View style={style.clearx}>
        <CompassIcon
            name='close-circle'
            color={changeOpacity(theme.centerChannelColor, 0.5)}
            size={20}
        />
    </View>
</TouchableWithoutFeedback>

<TapGestureHandler
    onHandlerStateChange={this.handleClear}
>
    <CompassIcon
        name='close-circle'
        color={changeOpacity(theme.centerChannelColor, 0.5)}
        style={style.clearx}
        size={20}
    />
</TapGestureHandler>


<TapGestureHandler
    onHandlerStateChange={this.handleClear}
>
    <View style={style.clearx}>
        <CompassIcon
            name='close-circle'
            color={changeOpacity(theme.centerChannelColor, 0.5)}
            size={20}
        />
    </View>
</TapGestureHandler>
```

**Comment by matthewbirtch** [comment]
> 0/5 I personally think it looks better with the "down chevron". Even though it is navigating to another screen, it still feels like a dropdown select box to me.

Maybe we remove the chevron and divider altogether for this type of input that triggers a new screen. It might not be needed at all

**Comment by enahum** [comment]
@mickmister is this ready for review? what about @matthewbirtch comments?

**Comment by mickmister** [comment]
@enahum @matthewbirtch Sorry for not following up here. I'm OOO and will be back in 2 weeks

**Comment by matthewbirtch** [comment]
> @enahum @matthewbirtch Sorry for not following up here. I'm OOO and will be back in 2 weeks

No worries @mickmister. I saw your note in the standup channel anyway. Enjoy your time off!

**Comment by mickmister** [comment]
> Maybe we remove the chevron and divider altogether for this type of input that triggers a new screen. It might not be needed at all

@matthewbirtch I've updated the PR to remove the chevron and divider. Here's how it looks now:



@enahum I'm really not a fan of using padding instead of the intended `hitSlop` prop solution you suggested to increase the tap area. As noted in https://github.com/mattermost/mattermost-mobile/pull/5398#discussion_r638151481, I'm having issues getting the `Touchable` components working properly with this component. Maybe there is something going on with the component structure (e.g. this component cannot be a descendent of this other component in order for that to work or something). Do you have any advice on how to proceed, because the padding solution is obviously not ideal.

**Comment by enahum** [comment]
> @enahum I'm really not a fan of using padding instead of the intended hitSlop prop solution you suggested to increase the tap area. As noted in #5398 (comment), I'm having issues getting the Touchable components working properly with this component

What issue are you having ? Perhaps you want us to have a pair programming session ?

**Comment by matthewbirtch** [comment]
> @matthewbirtch I've updated the PR to remove the chevron and divider. Here's how it looks now:

Thanks @mickmister. I think this is fine. Can we move the clear `x` icons a bit more to the right (something around 12-16 points to the right)? They look a little too far to the left now.

**Comment by mickmister** [comment]
> What issue are you having ? Perhaps you want us to have a pair programming session ?

@enahum Sure I'll DM you

**Comment by mickmister** [comment]
@DHaussermann Build is passing now

**Review by DHaussermann — APPROVED** [review]
Tested and passed 
- Clearing options work as expected  
- Clearing works for all applicable filed types (used full form modal)
- Tested on Android and iOS

LGTM!

@mickmister note that regression testing did find an issue with clearing multi-selects but as discussed, this seems unrelated to your PR. 

Created https://mattermost.atlassian.net/browse/MM-39866 to address separately.

**Comment by mickmister** [comment]
@DHaussermann I'm thinking the issue you pointed out _is_ caused by this PR. I'll look into the issue

**Comment by amyblais** [comment]
Thank you!

**Comment by enahum** [comment]
@mickmister can this be merged?

**Comment by mickmister** [comment]
@DHaussermann I've fixed the issue you pointed out above. Please take another look when you have the chance 🙂

**Comment by enahum** [comment]
@DHaussermann is this still on your radar?

**Comment by mickmister** [comment]
@DHaussermann gentle ping to review this

**Comment by DHaussermann** [comment]
This may still be testable with the test-app if I can get builds to deploy. I will attempt to setup a test for this shortly and let you know.

**Comment by DHaussermann** [comment]
@mickmister I have a setup in place for mobile PRs now and would love to close this out.
When you have a chance - can you please see if you can solve this build issue so I get a test build?

**Comment by mickmister** [comment]
@DHaussermann This should be ready to test

**Comment by DHaussermann** [comment]
@mickmister This is tested and passed on Andorid. 

I'm seeing something odd about the iOS test build generated here. Instead of a PR build based on `v1.55.0` as is the case for Android - iOS is giving me a build `v2.0.0` which is not showing me any Apps related binding in the input box. Thoughts?

**Comment by mickmister** [comment]
@DHaussermann The issue described in your comment seems unrelated to this PR. Let's have a followup discussion on what may be happening with the iOS app.

**Pull request merged**